openfe/protocols/restraint_utils/geometry/boresch.py

```diff
@@ -0,0 +1,949 @@
+                        # Get the sum of the average distances (dsum_avgs)
+                        # for all the host_atom_pool atoms
+                        distance1_avgs = np.array(
+                            [d.mean() for d in h2_eval.results.distances1]
+                        )
+                        distance2_avgs = np.array(
+                            [d.mean() for d in h2_eval.results.distances2]
+                        )
+                        dsum_avgs = distance1_avgs + distance2_avgs
+
+                        # Now filter by validity as H2 atom
+                        h2_dsum_avgs = [
+                            (idx, val) for idx, val in enumerate(dsum_avgs)
+                            if h2_eval.results.valid[idx]
+                        ]
+
+                        # Get the index of the H2 atom with the lowest
+                        # average distance
+                        k = sorted(h2_dsum_avgs, key=lambda x: x[1])[0][0]
+
+                        return list(host_atom_pool.atoms[[i, j, k]].ix)
+    return None
+
+
+def _get_restraint_distances(
+    atomgroup: mda.AtomGroup
+) -> tuple[unit.Quantity]:
+    """
+    Get the bond, angle, and dihedral distances for an input atomgroup
+    defining the six atoms for a Boresch-like restraint.
+
+    The atoms must be in the order of H0, H1, H2, G0, G1, G2.
+
+    Parameters
+    ----------
+    atomgroup : mda.AtomGroup
+      An AtomGroup defining the restrained atoms in order.
+
+    Returns
+    -------
+    bond : unit.Quantity
+      The H0-G0 bond value.
+    angle1 : unit.Quantity
+      The H1-H0-G0 angle value.
+    angle2 : unit.Quantity
+      The H0-G0-G1 angle value.
+    dihed1 : unit.Quantity
+      The H2-H1-H0-G0 dihedral value.
+    dihed2 : unit.Quantity
+      The H1-H0-G0-G1 dihedral value.
+    dihed3 : unit.Quantity
+      The H0-G0-G1-G2 dihedral value.
+    """
+
+    bond = calc_bonds(
+        atomgroup.atoms[0].position,
+        atomgroup.atoms[3],
+        box=atomgroup.dimensions
+    )
+
+    angles = []
+    for idx_set in [[1, 0, 3], [0, 3, 4]]:
+        angle = calc_angles(
+            atomgroup.atoms[idx_set[0]].position,
+            atomgroup.atoms[idx_set[1]].position,
+            atomgroup.atoms[idx_set[2]].position,
+            box=atomgroup.dimensions,
+        )
+        angles.append(angle * unit.radians)
+
+    dihedrals = []
+    for idx_set in [[2, 1, 0, 3], [1, 0, 3, 4], [0, 3, 4, 5]]:
+        dihed = calc_dihedrals(
+            atomgroup.atoms[idx_set[0]].position,
+            atomgroup.atoms[idx_set[1]].position,
+            atomgroup.atoms[idx_set[2]].position,
+            atomgroup.atoms[idx_set[3]].position,
+            box=atomgroup.dimensions,
+        )
+        dihedrals.append(dihed * unit.radians)
+
+    return bond, angles[0], angles[1], dihedrals[0], dihedrals[1], dihedrals[2]
+
+
+def find_boresch_restraint(
+    topology: Union[str, pathlib.Path, openmm.app.Topology],
```

Comment:
Switch this to just take in a Universe? It would make it easier to deal with the single PDB case.